src/test/java/io/confluent/examples/streams/CustomStreamTableJoinIntegrationTest.java

```diff
@@ -0,0 +1,538 @@
+          final Long tableValue = tableStore.get(key);
+          if (tableValue != null) {
+            final KeyValue<String, Pair<Double, Long>> joinRecord = KeyValue.pair(key, new Pair<>(value, tableValue));
+            LOG.info("Table data available for key {}, sending fully populated join message {}", key, joinRecord);
+            return joinRecord;
+          } else {
+            LOG.info("Table data unavailable for key {}, buffering stream record ({}, {}) temporarily", key, key, value);
+            streamBufferStore.put(key, new Pair<>(value, Instant.ofEpochMilli(streamRecordTimestamp)));
+            return null;
+          }
+        }
+
+        private void punctuate(final long timestamp) {
+          LOG.info("Punctuating @ timestamp {}", timestamp);
+          sendAndPurgeAnyWaitingRecordsThatHaveExceededWaitTime(timestamp);
+        }
+
+        private void sendAndPurgeAnyWaitingRecordsThatHaveExceededWaitTime(final long currentStreamTime) {
+          try (KeyValueIterator<String, Pair<Double, Instant>> iterator = streamBufferStore.all()) {
+            while (iterator.hasNext()) {
+              final KeyValue<String, Pair<Double, Instant>> record = iterator.next();
+              LOG.info("Checking waiting stream record ({}, {}) with timestamp {}", record.key, record.value.x,
+                record.value.y.toEpochMilli());
+              if (waitTimeExpired(record.value.y, currentStreamTime)) {
+                final Pair<Double, Long> joinedValue = new Pair<>(record.value.x, null);
+                LOG.info("Wait time for stream record ({}, {}) expired, force-forwarding now as join message " +
+                  "({}, ({}, {}))", record.key, record.value.x, record.key, joinedValue.x, joinedValue.y);
+                context.forward(record.key, joinedValue);
+                streamBufferStore.delete(record.key);
+              }
+            }
+          }
+        }
+
+        private boolean waitTimeExpired(final Instant recordTimestamp, final long currentStreamTime) {
+          return Duration.between(recordTimestamp, Instant.ofEpochMilli(currentStreamTime))
+            .compareTo(approxMaxWaitTimePerRecordForTableData) > 0;
+        }
+
+        @Override
+        public void close() {
+        }
+
+      };
+    }
+
+  }
+
+  /**
+   * Implements table-side triggering of join output.  This is but one example for such custom join semantics -- feel
+   * free to modify this example to match your own needs.
+   *
+   * For every <i>observed</i> record arriving at its upstream table, this transformer will check for a buffered (i.e.,
+   * not yet joined) record on the stream side.  If there is a match, then the transformer will produce a fully
+   * populated join output message -- which is the desired table-side triggering behavior.  If there is no match, then
+   * the transformer will effectively do nothing (it will actually send a tombstone record).
```

Comment:
> The output is a KStream [...]
No, the output is a `KTable`. Remember, the output of `KTable#transformValues()` is another `KTable`.

(We do convert the output KTable later on into a stream so we can filter out the tombstone records, but that's different!)

Reply:
Ah. This is `KTable#transformValues()` -- however, this makes me think if it would be better to do `KTable#toString()#transformValues()`, because the output of a stream-table join should be a `KStream`? This would also allow to use `flatTransformValues()` later, that is only added to `KStream` -- this way, the downstream `filter()` can be avoided.

Reply:
I suppose you mean `KTable#toStream()#transformValues()`, not `KTable#toString()#transformValues()` ;-)
 
Regarding your suggestion: I don't think this really makes a difference. The key UX issue IMHO is that we don't have an explicit `.join()` operation anywhere in the code because the joining happens inside the various `Transformers`, and that's why the last "real" step in the topology is the merging of the transformed input stream and the stream of the transformed table.  Switching to `KTable#toStream()#transformValues()` doesn't change this.

```java
    final KStream<String, Pair<Double, Long>> joined =
        transformedStream.merge(transformedTable.toStream().filter((k, v) -> v != null));
```

Reply:
Yes, you still need to `merge()`, however, you don't need to `filter()` using upcoming `KStream#flatValueTransform()`.